pkg/maps/ipmasq/ipmasq.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ipmasq
+
+import (
+	"fmt"
+	"net"
+	"unsafe"
+
+	"github.com/cilium/cilium/common/types"
+	"github.com/cilium/cilium/pkg/bpf"
+)
+
+const (
+	MapName    = "cilium_ipmasq_v4"
+	maxEntries = 16384
+)
+
+// +k8s:deepcopy-gen=true
+// +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf.MapKey
+type Key4 struct {
+	PrefixLen uint32
+	Address   types.IPv4
+}
+
+func (k *Key4) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(k) }
+func (k *Key4) NewValue() bpf.MapValue    { return &Value{} }
+func (k *Key4) String() string            { return fmt.Sprintf("%s", k.Address) }
+
+// +k8s:deepcopy-gen=true
+// +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf.MapValue
+type Value struct {
+	Pad uint8 // not used
+}
+
+func (v *Value) GetValuePtr() unsafe.Pointer { return unsafe.Pointer(v) }
+func (v *Value) String() string              { return "" }
+
+var IPMasq4Map = bpf.NewMap(
+	MapName,
+	bpf.MapTypeLPMTrie,
+	&Key4{}, int(unsafe.Sizeof(Key4{})),
+	&Value{}, int(unsafe.Sizeof(Value{})),
+	maxEntries,
+	bpf.BPF_F_NO_PREALLOC, 0,
+	bpf.ConvertKeyValue,
+).WithCache()
+
+type IPMasqBPFMap struct{}
+
+func (*IPMasqBPFMap) Update(cidr net.IPNet) error {
+	return IPMasq4Map.Update(key(cidr), &Value{})
+}
+
+func (*IPMasqBPFMap) Delete(cidr net.IPNet) error {
+	return IPMasq4Map.Delete(key(cidr))
+}
+
+func (*IPMasqBPFMap) Dump() ([]net.IPNet, error) {
+	cidrs := []net.IPNet{}
+	if err := IPMasq4Map.DumpWithCallback(
+		func(key bpf.MapKey, value bpf.MapValue) {
+			cidrs = append(cidrs, keyToIPNet(key.(*Key4)))
+		}); err != nil {
+		return nil, err
+	}
+	return cidrs, nil
+}
+
+func key(cidr net.IPNet) *Key4 {
+	ones, _ := cidr.Mask.Size()
+	key := &Key4{PrefixLen: uint32(ones)}
+	copy(key.Address[:], cidr.IP.To4())
```

Comment:
Is anything validating, for instance from the config file parsing -> `Update()` -> `key()`, that `cidr` here is IPv4 and that the `To4()` call here won't return `nil`?

Reply:
Good point, I've added the validation to the caller (`pkg/ipmasq/ipmasq.go:(c *Ipnet) UnmarshalYAML()`), so that errors could be reported earlier.